### PR TITLE
Switch from Flask to Starlette

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: mergify_engine/sub_utils.py [hunk 1]
```diff
@@ -121,9 +122,10 @@ def _retrieve_subscription_from_cache(r, installation_id):
...
 
-def save_subscription_to_cache(r, installation_id, sub):
+async def save_subscription_to_cache(installation_id, sub):
+    r = await utils.get_aioredis_for_cache()
     encrypted = _encrypt(sub)
-    r.set("subscription-cache-%s" % installation_id, encrypted, ex=3600)
+    await r.setex("subscription-cache-%s" % installation_id, 3600, encrypted)
 
 
 def get_subscription(r, installation_id):
```
Comment (jd) [inline mergify_engine/sub_utils.py:131]: TODO async I guess?

File: mergify_engine/tests/functional/test_engine.py [hunk 2]
```diff
@@ -752,7 +752,9 @@ def test_merge_smart_strict(self):
                 "Content-type": "application/json",
             },
         )
-        self.assertEqual(r.json, {"mergifyio-testing/%s" % self.name: {"master": [2]}})
+        self.assertEqual(
+            r.json(), {"mergifyio-testing/%s" % self.name: {"master": [2]}}
+        )
```
Comment (jd) [inline mergify_engine/tests/functional/test_engine.py:757]: feel free to replace with `assert` while you're at it

File: mergify_engine/tests/unit/test_badges.py [hunk 3]
```diff
@@ -14,29 +14,36 @@
...
 
 def test_badge_redirect():
-    with web.app.test_request_context("/"):
-        reply = web._get_badge_url("mergifyio", "mergify-engine", "png")
-        assert reply.status_code == 302
+    with testclient.TestClient(web.app) as client:
+        reply = client.get(
+            "/badges/mergifyio/mergify-engine.png", allow_redirects=False
+        )
+        assert reply.status_code == 307
```
Comment (jd) [inline mergify_engine/tests/unit/test_badges.py:27]: It sounds to me that 302 was the right choice, no? Why changing, why here, why now? 🤔

File: mergify_engine/web.py [hunk 4]
```diff
@@ -62,93 +66,128 @@ def authentification():  # pragma: no cover
...
 
 
-@app.route("/badges/<owner>/<repo>")
-def badge(owner, repo):
-    return flask.redirect(f"https://dashboard.mergify.io/badges/{owner}/{repo}")
+@app.get("/badges/{owner}/{repo}")
+async def badge(owner, repo):
+    return responses.RedirectResponse(
+        url=f"https://dashboard.mergify.io/badges/{owner}/{repo}"
+    )
```
Comment (jd) [inline mergify_engine/web.py:98]:
```suggestion
        status_code=302,
    )
```

File: mergify_engine/web.py [hunk 5]
```diff
@@ -202,157 +241,180 @@ def voluptuous_error(error):
...
-        args=[event_type, event_id, data], countdown=30
+    loop = asyncio.get_running_loop()
+    await loop.run_in_executor(
+        None,
+        functools.partial(
+            github_events.job_filter_and_dispatch.s(
+                event_type, event_id, data
+            ).apply_async,
+            countdown=30,
+        ),
```
Comment (jd) [inline mergify_engine/web.py:338]: IIUC this is just a hack to run the function concurrently, right?

Comment (sileht) [reply mergify_engine/web.py:338]: Yes, that should be temporary as the rest will switch to async too.

File: setup.cfg [hunk 6]
```diff
@@ -25,12 +25,16 @@ install_requires =
     daiquiri>=2.0.0
     datadog
     python-dotenv
-    flask
-    flask_cors
+    fastapi
+    python-multipart
+    ujson
```
[the remark this answers is not in the document]
Comment (sileht) [reply setup.cfg:30]: their extras sucks:
```
{'all': ['requests',
         'aiofiles',
         'jinja2',
         'python-multipart',
         'itsdangerous',
         'pyyaml',
         'graphene',
         'ujson',
         'orjson',
         'email_validator',
         'uvicorn',
         'async_exit_stack',
         'async_generator'],
```

I double checked we just need `uvicorn`

File: mergify_engine/tests/unit/test_badges.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/tests/unit/test_badges.py:27]: Right, the default between flask and starlette differ, I fixed it.

File: mergify_engine/sub_utils.py (hunk 1 above)
Comment (sileht) [reply mergify_engine/sub_utils.py:131]: A lot are TODO for async, method will migrate slowly :)

**Pull request merged**